store/src/java/com/zimbra/cs/account/ldap/LdapProvisioning.java

```diff
@@ -6219,6 +6240,42 @@ private void checkPasswordStrength(String password, Account acct, Cos cos, ZMuta
         }
     }
 
+  // validate that user is not using personal data in his password
+  private void validatePasswordEntropy(String password, Account acct)
+      throws AccountServiceException {
+    String s_name = acct.getAttr(Provisioning.A_sn) == null ? "" : acct.getAttr(Provisioning.A_sn);
+    String c_name = acct.getAttr(Provisioning.A_cn) == null ? "" : acct.getAttr(Provisioning.A_cn);
+    String initials = acct.getAttr(Provisioning.A_initials) == null ? "" : acct.getAttr(Provisioning.A_initials); //could be null
+    String email = acct.getAttr(Provisioning.A_mail);
+
+    String[] fullNameExploded = c_name.toLowerCase().split(" ");
+    String[] emailExploded = email.split("[@._]");
+    String[] cleanEmailExploded =
+        Arrays.copyOf(emailExploded, emailExploded.length - 1); // remove the top level domain(TLD)
+    String passwordLower = password.toLowerCase();
+
+    String[] personalDataImploded = {
+      s_name.toLowerCase(),
+      c_name.toLowerCase(),
+      c_name.replace(" ", "").toLowerCase(),
+      initials.toLowerCase()
+    };
+
+    String[] dictionary =
+        Stream.of(personalDataImploded, cleanEmailExploded, fullNameExploded)
+            .flatMap(Stream::of)
+            .toArray(String[]::new);
+
+    for (String part : dictionary) {
+      if (part.length() > 2 && (part.contains(passwordLower) || passwordLower.contains(part))) {
+        throw AccountServiceException.INVALID_PASSWORD(
+            "password contains username or other personal data: " + part,
+            new Argument(
+                Provisioning.A_zimbraBlockPersonalDataInPasswordEnabled, part, Argument.Type.STR));
+      }
+    }
```

Comment:
You can execute all this code in the same stream for better performances:
```suggestion
    Optional<String> matchedSensitivePart =
        Stream.of(personalDataImploded, cleanEmailExploded, fullNameExploded)
            .flatMap(Stream::of)
            .parallel()
            .anyMatch(part -> part.length() > 2 && (part.contains(passwordLower) || passwordLower.contains(part))))
            .findAny();

    if (matchedSensitivePart.isPresent()) {
        throw AccountServiceException.INVALID_PASSWORD(
            "password contains username or other personal data: " + part,
            new Argument(
                Provisioning.A_zimbraBlockPersonalDataInPasswordEnabled, part, Argument.Type.STR));
    }
```